packages/flutter/lib/src/cupertino/text_field.dart

```diff
@@ -1145,59 +1148,69 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with Restoratio
     return ValueListenableBuilder<TextEditingValue>(
       valueListenable: _effectiveController,
       child: editableText,
-      builder: (BuildContext context, TextEditingValue? text, Widget? child) {
+      builder: (BuildContext context, TextEditingValue text, Widget? child) {
+        final bool hasText = text.text.isNotEmpty;
+        final String? placeholderText = widget.placeholder;
+        final Widget? placeholder = placeholderText == null
+          ? null
+          // Make the placeholder invisible when hasText is true.
+          : Visibility(
+              maintainAnimation: true,
+              maintainSize: true,
+              maintainState: true,
+              visible: !hasText,
+              child: SizedBox(
+                width: double.infinity,
+                child: Padding(
+                  padding: widget.padding,
+                  child: Text(
+                    placeholderText,
+                    // This is to make sure the text field is always tall enough
+                    // to accommodate the first line of the placeholder, so the
+                    // text does not shrink vertically as you type (however in
+                    // rare circumstances, the height may still change when
+                    // there's no placeholder text).
+                    maxLines: hasText ? 1 : widget.maxLines,
+                    overflow: placeholderStyle.overflow,
+                    style: placeholderStyle,
+                    textAlign: widget.textAlign,
+                  ),
+                ),
+              ),
+          );
+
+        final Widget? prefixWidget = _shouldShowAttachment(attachment: widget.prefixMode, hasText: hasText) ? widget.prefix : null;
+
+        // Show user specified suffix if applicable and fall back to clear button.
+        final bool showUserSuffix = _shouldShowAttachment(attachment: widget.suffixMode, hasText: hasText);
+        final bool showClearButton = _shouldShowAttachment(attachment: widget.clearButtonMode, hasText: hasText);
+        final Widget? suffixWidget = switch ((showUserSuffix, showClearButton)) {
+          (false, false) => null,
+          (true, false) => widget.suffix,
+          (true, true) => widget.suffix ?? _buildClearButton(),
+          (false, true) => _buildClearButton(),
+        };
         return Row(children: <Widget>[
           // Insert a prefix at the front if the prefix visibility mode matches
           // the current text state.
-          if (_showPrefixWidget(text!)) widget.prefix!,
+          if (prefixWidget != null) prefixWidget,
           // In the middle part, stack the placeholder on top of the main EditableText
           // if needed.
           Expanded(
             child: Stack(
+              // Ideally this should be baseline aligned. However that comes at
```

Comment:
Since this change, the placeholder is vertically centered on multi-line textfield instead of being on the first line, because alignment is set to center instead of default topLeft. Is this intended? Not sure how to read the comment above.

Reply:
Yeah it should be baseline aligned with the input text instead of center aligned. I'll try to fix that.